README.md
# Python Module for Ancient Indian Mathematics and Astronomy

- [X] 1. Aryabhata Cryptographic Calculator
    1. Calculate values from cryptographic words.
    1. Generate cryptographic words from numbers.
- [X] 2. Square Root When the Smaller Square Root is Known (Using Bodhaayana Approximation Method)
- [X] 3. Square Root (Bhaskara Method)
- [x] 4. Cube Root (Bhaskara Method)
- [x] 5. Solar Eclipse Calculator (Improved Siddhantic Procedure)
- [x] 6. Lunar Eclipse Calculator (Improved Siddhantic Procedure)
---
### Module Link
Python Module for Ancient Indian Mathematics and Astronomy
[AncIndMatAst](https://pypi.org/project/AncIndMatAst/)

### Installation
```
pip install AncIndMatAst
```
### Aryabhata Cryptographic Calculator
![Aryabhata](images/Aryabhata.png)
##### Shlok
```
वर्गात्राणि वर्गेऽवर्गेऽवर्गाताराणी का मौ यः ।
खद्विनवके स्वरा नव वर्गेऽवर्गे नवान्त्यवर्गे वा ॥
```

###### Table 1
Varg Consonants From `क्` to `म्` denote from `1` to `25`

|वर्ग:  | क | ख | ग | घ् | ङ् | च् | छ् | ज्  | झ् | ञ् | ट् | ठ् | ड् |
|-----|----|---|---|---|----|---|----|---|---|----|---|---|----|
|व्यञ्जन | 1 | 2 | 3 | 4 | 5   | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 |
|     | ढ् | ण् | त् | थ् | द् | ध् | न् | प् | फ् | ब् | भ् | म् |
|     |14 |15  | 16| 17 | 18 | 19 | 20 | 21 | 22 |23 | 24 | 25 |

###### Table 2
Awarg consonants from `य्` to `ह्` denote from `3` to `10`

| अवर्गः | य् | र् | ल् | व् | श् | ष | स | ह |
|-----|----|---|---|---|----|---|----|---|
| व्यञ्जन | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 |

###### Table 3
Vowels from `अ` to `औ` denote place-values from <code>10<sup>0</sup></code> to <code>10<sup>17</sup></code>

| स्वरः | अ | इ | उ | ऋ | लृ | ए | ऐ | ओ | औ |
|-----|----|---|---|---|----|---|----|---|---|
| वर्ग: | 10^0^ | 10^2^ | 10^4^ | 10^6^ | 10^8^ | 10^10^ | 10^12^ | 10^14^ | 10^16^ |
| अवर्गः | 10^1^ | 10^3^ | 10^5^ | 10^7^ | 10^9^ | 10^11^ | 10^13^ | 10^15^ | 10^17^ |
----

ढुविघ्व is the Aryabhatiya Alphabetical Numeral for the number of revolutions of Saturn in 43,20,000 years, and it is 1,46,564.
###### Explanation
ढुङ्विध्व = ( ढ उ ) + (ङ् इ) + (व् इ) + (घ् अ) + (व् अ) 
>= (14×10^4^ ) + (5×10^2^) + (6 ×10^3^) + (4 x 1 ) + (6 x 10 )
>=1,46,564.

---
#### Decode
##### Code
```python
from AncIndMatAst import AC
word="जषबिखुछृ"
value=AC.decode(word)
print(f"Value of {word} = {value}")
```
##### Output
```
Value of जषबिखुछृ = 7022388
```
#### Encode
##### Code
```python
from AncIndMatAst import AC
data=AC.encode(364224)
print(data.get("sabd"))
print(data.get("length"))
print(data.get("allSabd"))
```
##### Output
```
भखिरिचुयु
120
['भखिरिचुयु', 'भखिरियुचु', 'भखिचुरियु', 'भखिचुयुरि', 'भखियुरिचु', 'भखियुचुरि', 'भरिखिचुयु', 'भरिखियुचु', 'भरिचुखियु', 'भरिचुयुखि', 'भरियुखिचु', 'भरियुचुखि', 'भचुखिरियु', 'भचुखियुरि', 'भचुरिखियु', 'भचुरियुखि', 'भचुयुखिरि', 'भचुयुरिखि', 'भयुखिरिचु', 'भयुखिचुरि', 'भयुरिखिचु', 'भयुरिचुखि', 'भयुचुखिरि', 'भयुचुरिखि', 'खिभरिचुयु', 'खिभरियुचु', 'खिभचुरियु', 'खिभचुयुरि', 'खिभयुरिचु', 'खिभयुचुरि', 'खिरिभचुयु', 'खिरिभयुचु', 'खिरिचुभयु', 'खिरिचुयुभ', 'खिरियुभचु', 'खिरियुचुभ', 'खिचुभरियु', 'खिचुभयुरि', 'खिचुरिभयु', 'खिचुरियुभ', 'खिचुयुभरि', 'खिचुयुरिभ', 'खियुभरिचु', 'खियुभचुरि', 'खियुरिभचु', 'खियुरिचुभ', 'खियुचुभरि', 'खियुचुरिभ', 'रिभखिचुयु', 'रिभखियुचु', 'रिभचुखियु', 'रिभचुयुखि', 'रिभयुखिचु', 'रिभयुचुखि', 'रिखिभचुयु', 'रिखिभयुचु', 'रिखिचुभयु', 'रिखिचुयुभ', 'रिखियुभचु', 'रिखियुचुभ', 'रिचुभखियु', 'रिचुभयुखि', 'रिचुखिभयु', 'रिचुखियुभ', 'रिचुयुभखि', 'रिचुयुखिभ', 'रियुभखिचु', 'रियुभचुखि', 'रियुखिभचु', 'रियुखिचुभ', 'रियुचुभखि', 'रियुचुखिभ', 'चुभखिरियु', 'चुभखियुरि', 'चुभरिखियु', 'चुभरियुखि', 'चुभयुखिरि', 'चुभयुरिखि', 'चुखिभरियु', 'चुखिभयुरि', 'चुखिरिभयु', 'चुखिरियुभ', 'चुखियुभरि', 'चुखियुरिभ', 'चुरिभखियु', 'चुरिभयुखि', 'चुरिखिभयु', 'चुरिखियुभ', 'चुरियुभखि', 'चुरियुखिभ', 'चुयुभखिरि', 'चुयुभरिखि', 'चुयुखिभरि', 'चुयुखिरिभ', 'चुयुरिभखि', 'चुयुरिखिभ', 'युभखिरिचु', 'युभखिचुरि', 'युभरिखिचु', 'युभरिचुखि', 'युभचुखिरि', 'युभचुरिखि', 'युखिभरिचु', 'युखिभचुरि', 'युखिरिभचु', 'युखिरिचुभ', 'युखिचुभरि', 'युखिचुरिभ', 'युरिभखिचु', 'युरिभचुखि', 'युरिखिभचु', 'युरिखिचुभ', 'युरिचुभखि', 'युरिचुखिभ', 'युचुभखिरि', 'युचुभरिखि', 'युचुखिभरि', 'युचुखिरिभ', 'युचुरिभखि', 'युचुरिखिभ']
Value of जषबिखुछृ = 7022388
```
---
### Bodhaayana Square Root (Approximation Method)
![Bodhaayana](images/Bodhaayana.png)
##### Shlok
```
प्रमाणं तृत्तीयेन वर्धयेत् तच्च चतुर्थेनात्म चतुस्त्रिम्शोनेन सविशेष:
```
##### Explanation
The measure is to be increased by its third and this (third) again by its own fourth less the thirty fourth part (of that fourth); this is (the value of) the diagonal of a square (whole side is the measure).

##### Code
```python
from AncIndMatAst import BA
num=200
sq=BA.squareRoot(num)
print("Square Root By Bodhaayana Approximation Method = ",sq)
```
##### Output
```
Square Root By Bodhaayana Approximation Method =  14.1421356237469
```
---
### Roots (Bhaskara Method)
![Bhaskara](images/Bhaskara.png)
### Square Root
###### Shlok
```
Tyaktvaamtyaat Vishamaat kritim
    Dvigunayenmulam same taddhrite 
Tyaktvaa labdhakritim tadaadya vishamaat 
    Labdham dvinighnam nyaset

Panktyaam panktihrite samenyavishamaat 
    Tyaktvaptavargam phalam 
Panktyaam taddvigumam nyasediti muhuh 
    Pankterdalam syaatpadam
```
###### Explanation
1. Start observing the given number from RHS
2. Put dots on the digits appearing at all the odd places (that is, 1, 3, 5, 7 etc. places) starting from RHS
3. Start processing the given number from LHS. 
4. Remember that the terms of (a+b)=a²+2ab+b² formula will be used to find the square root 
5. Select the leftmost set of digits having the dot and call it X1.
6. Identify a value of a (where a=1,2,3,4,5,6,7,8,9) which is the biggest value that can be subtracted from XI.
7. Keep the corresponding value of 'a' in the result location.
8. Subtract that a from XI and call the resultant value X2.
9. Append the next digit of the given number to X2 for continuing the processing. Call it as X3. 
10. Divide X3 by 2*a. Identify the quotient as 'b' (Maximum value of b can be 9).
11. Subtract 2*a*b from X3 and call the resultant value X4
12. Append the next digit of the given number to X4 for continuing the processing. Call
it as X5. 
13. Subtract b³ from X5 and call the resultant value X6,
14. If any of the above two subtractions give negative value, reduce the value of bby one and revert back to step 11.
15. Concatenate the values of 'a' and 'b'. Call it as 'ab'. 
16. This value is the square root of the group of digits processed so far.
17. If some more digits are still pending for processing in the given number, take the present value of ab as new 'a'. Take the value of X6 as X2. Repeat the process from step 9.

##### Code
```python
from AncIndMatAst import BM
num=144
sq=BM.squareRoot(num)
print("Square Root By Bhaskara Method = ",sq)
```
##### Output
```
Square Root By Bhaskara Method =  12
```
---
### Cube Root
###### Shlok
```
Adyam ghanasthanamathaaghane dve 
    Punastathantyaat ghanto visodhya
Ghanam prithakstham padamasya krityan
    Trighnyaa tadaadyam vibhajet phalam tu

Panktyaam nyaset tatkritimantyanighneem
    Trighneem tyajet tatprathamaat phalasya
Ghanam tadaadyaat ghanamulamevam
Panktirbhavedevamatah punasca
```
---
###### Explanation
1. Start observing the given number from RHS
2. Put dots on the digits appearing at 1st , 4th , 7th , 9th etc. places starting from RHS
3. Start processing the given number from LHS
4. Remember that the terms of (a + b)3 = a ^ 3 + 3a ^ 2 * b + 3a * b ^ 2 + b3 formula will be used to find the cube root
5. Select the leftmost set of digits having the dot and call it XI.
6. Identify a value of a ^ 3 where a (1,2,3,4,5,6,7,8,9) which is the biggest value that can be subtracted from XI.
7. keep the corresponding value of 'a' in the result location. 8. Subtract that a ^ 2 from X1 and call the resultant value X2.
9. Append the next digit of the given number to X2 for continuing the processing. Call it as X3.
10. Divide X3 by 3^ * a ^ 2 Identify the quotient as "b.(Maximum value of b can be 9).
11. Subtract 3 *a^ 2* b from X3 and call the resultant value X4
12. Append the next digit of the given number to X4 for continuing the processing. Call it as XS.
13. Subtract 3 a b from X5 and call the resultant value X6.
14. Append the next digit of the given number to X6 for continuing the processing. Call it
as X7.
15. Subtract b' from X7 and call the resultant value X8.
16. If any of the above three subtractions give negative value, reduce the value of b by one and revert back to step 11.
17. Concatenate the values of "a" and "b". Call it as 'ab'.
18. This value will be the cube root of the group of digits processed so far.
19. If some more digits are still pending for processing in the given number, take the present value of ab as new 'a' Take the value of X8 as X2. Repeat the process from step 9

###### Illustration
```
To Find the cube root of 9261:
1. Start observing the given number from RHS
2. Put dots on the digits appearing at 1st and 4th places starting from RHS, viz., or 1 and 9.
3. Start processing the given number from LHS
4. Remember that the terms of (a + b)^3 = a ^ 3 + 3a ^ 2 * b + 3a * b ^ 2 + b ^ 1 formula will be used in find the cube root
5. Select the leftmost set of digits having the dot and call it XI.
6. X1=9
7. Identity a value of a ^ 3 * (- 8) which is the biggest value that can be subtracted from XI.
8. keep the corresponding value of ^ * a' (=2) in the result location.
9 Subtract that a from X1 and call the resultant value X2.
10. X2 = 9 - 8 = 1
11. Append the next digit of the given number to X2 for continuing the processing Call it as X3.
12. X3 = 12
13. Divide X3 by 3*a ^ 2 (=12)
14. Identify the quotient as "b".
15. b=1
16. Subtract 3 ab from X3 and call the resultant value X4
17. X4 = 12 - 12 = 0
18. Append the next digit of the given number to X4 for continuing the processing Cal it as X5.
19. X5 = 06
20. Subtract 3^ * a^ * b ^ 2 from X5 and call the resultant value X6.
21. X6=06-6-0
22. Append the next digit of the given minber to X6 for continuing the processing. Call it us XT.
23. X7=01
24. Subtract b' from X7 and call the resultant value X8
25. X8=01-1-0
26. Concatenate the values of 'a' and 'b. Call it as ab 27, ab21
28. This value 21 is the cube root of the given number 9261
Result = 21
```

##### Code
```python
from AncIndMatAst import BM
num=175616
sq=BM.cubeRoot(num)
print("Cube Root By Bhaskara Method = ",sq)
```
##### Output
```
Cube Root By Bhaskara Method =  56
```

---
### Siddhantic Procedure for Lunar and Solar Eclipse
![Siddhantic](images/Sidddhantic.png)
#### Lunar Eclipse
![Lunar Eclipse](images/Lunar.png)
###### Code
```python
from AncIndMatAst import Eclipse
date = (2006, 10, 7)
time = (24,12,0)
trueMoon=(321,3,27)
trueRahu=(331,21,11)
Eclipse.LunarEclipse(date,time,trueMoon,trueRahu)
```
###### Output
```
Lunar Eclipse Detail of 07 October 2006
Eclipse Type :  Partial
Magnitude of Eclipse :  0.18
Timings of Eclipse :
Beginnig of Eclipse :  23h 35m 23s
Middle of Eclipse :  24h 21m 24s
End of Eclipse :  25h 7m 25s
```

#### Solar Eclipse
![Solar Eclipse](images/Solar.png)
###### Code
```python
from AncIndMatAst import Eclipse
date = (2008, 8, 1)
iONMoon=(15,42,0)
trueMoon=(105,33,0)
trueRahu=(294,37,26)
Eclipse.SolarEclipse(date,iONMoon,trueMoon,trueRahu)
```
###### Output
```
Solar Eclipse Detail of 01 August 2008
Eclipse Type :  Total
Magnitude of Eclipse :  1.34
Timings of Solar Eclipse :
Beginnig of Eclipse :  13h 31m 13s
Beginnig of Totality :  14h 45m 14s
Middle of Eclipse :  15h 50m 15s
End of Totality :  16h 55m 16s
End of Eclipse :  18h 8m 18s
```

CHANGELOG.txt
Change Log
==========

0.0.1 (11/12/2023)
-------------------
- First Release

test.py
from AncIndMatAst import BM,AC,BA,Eclipse


data=AC.encode(364224)
print(data.get("sabd"))
print(data.get("length"))
print(data.get("allSabd"))


# ढुङ्विध्व
# ढुङिविघव
# खिच्युभ
# भदिलझुनुखृ
# ख्युघृ
# जषबिखुछृ
# सुगुशिथृन
# चयगियिङुशुछृलृ
word="जषबिखुछृ"
value=AC.decode(word)
print(f"Value of {word} = {value}")




# Bhaskara Method 
# Square Root 
num=144
sq=BM.squareRoot(num)
print("Square Root By Bhaskara Method = ",sq)

# Cube Root 
num=175616
sq=BM.cubeRoot(num)
print("Cube Root By Bhaskara Method = ",sq)


# Bodhaayana Approximation Method 
num=200
sq=BA.squareRoot(num)
print("Square Root By Bodhaayana Approximation Method = ",sq)

# Sidhantic Procedure For Eclipse 
# Lunar Eclipse
# Example 1
date = (2006, 10, 7)
time = (24,12,0)
trueMoon=(321,3,27)
trueRahu=(331,21,11)
Eclipse.LunarEclipse(date,time,trueMoon,trueRahu)

# Solar Eclipse 
date = (2008, 8, 1)
iONMoon=(15,42,0)
trueMoon=(105,33,0)
trueRahu=(294,37,26)

Eclipse.SolarEclipse(date,iONMoon,trueMoon,trueRahu)

AncIndMatAst/BM.py
def squareRoot(x):
    x_s=str(x)
    l=len(x_s)
    il=l%2
    if il==0:
        il=2
    x1=int(x_s[0:il])
    p=il
    for i in range(1,11):
        if(x1<i*i):
            a=i-1
            break
    res=a
    x2=x1-a*a
    if p==l:
        return res
    # while all digits are not completed
    while True:
        x3=x2*10+int(x_s[p])
        b=x3//(2*a)+1
        # while not subtraction is not positive 
        while True:
            b=b-1
            x4=x3-(2*a*b)
            x5=x4*10+int(x_s[p+1])
            x6=x5-b*b
            if(x4>=0 and x6>=0):
                res=res*10+b
                p=p+2
                break
        if(p<l-1):
            a=res
            x2=x6
        else:
            break
    return res


def cubeRoot(x,showSteps=0):
    x_s=str(x)
    l=len(x_s)
    il=l%3
    if il==0:
        il=3
    x1=int(x_s[0:il])
    if(showSteps):
        print(f"X1 = {x1}")
    p=il
    for i in range(1,11):
        if(x1<i*i*i):
            a=i-1
            break
    if(showSteps):
        print(f"a = {a}")
    res=a
    x2=x1-a*a*a
    if(showSteps):
        print(f"X2 = {x2}")
    if p==l:
        return res
    # while all digits are not completed
    while True:
        x3=x2*10+int(x_s[p])
        if(showSteps):
            print(f"X3 = {x3}")
        b=x3//(3*a*a)+1
        # while not subtraction is not positive 
        while True:
            b=b-1
            if(showSteps):
                print(f"b = {b}")
            x4=x3-(3*a*a*b)
            if(showSteps):
                print(f"X4 = {x4}")
            x5=x4*10+int(x_s[p+1])
            if(showSteps):
                print(f"X5 = {x5}")
            x6=x5-3*a*b*b
            if(showSteps):
                print(f"X6 = {x6}")
            x7=x6*10+int(x_s[p+2])
            if(showSteps):
                print(f"X7 = {x7}")
            x8=x7-b*b*b
            if(showSteps):
                print(f"X8 = {x8}")
            if(x4>=0 and x6>=0 and x8>=0):
                res=res*10+b
                p=p+3
                break
        if(p<l-1):
            a=res
            x2=x8
        else:
            break
    return res

# print(f"Result = {squareRoot(200)}")

AncIndMatAst/Eclipse.py
import math,datetime
epoch = datetime.datetime(2000, 1, 1)

# To convert angles into second 
def angleToSec(Deg):
    h,m,s=Deg
    return h*60*60+m*60+s

# To convert seconds into angle 
def secToAngle(sec):
    neg=False
    if(sec<0):
        neg=True
        sec=-1*sec
    h=sec//3600
    if neg:
        h=-1*h
    sec=sec%3600
    m=sec//60
    sec=sec%60
    s=round(sec)
    return (h,m,s)

# To convert vinadi to second
def vinadiToSec(nadi):
    return nadi*24

# To convert nadi into second 
def nadiToSec(nadi):
    return nadi*24*60

# To convert naligai into second 
def naligaiToSec(naligai):
    return (naligai*12*3600)//30

# To format time while printing
def printTime(sec):
    time=secToAngle(sec)
    return f"{round(time[0])}h {round(time[1])}m {round(time[0])}s"

# Lunar Eclipse Calculator 
def LunarEclipse(date,time,trueMoon,Rahu):
    """
    Calculate Lunar Eclipse 
    Parameters :
    date => Date to check for eclipse in (Y,M,D)
    time => Instant of opposition
    trueMoon => True Longitude of Moon
    trueRahu => True Longitude of Rahu

    Output:
    Eclipse Type
    Eclipse Magnitude
    If Eclipse Occurs
    Eclipse Beginning Time
    Eclipse Middle Time
    Eclipse End Time
    """
    date=datetime.datetime(*date)
    trueMoonInSec=angleToSec(trueMoon)
    trueRahuInSec=angleToSec(Rahu)

    diff=trueMoonInSec-trueRahuInSec

    # initially 308 but better result 
    beta=angleToSec((0,318,0))*math.sin(math.radians(diff/3600))

    daysDiff=(date-epoch).days
    gm=(134.9633964+13.06499295*daysDiff)%360
    gs=(357.5291092+0.985600231*daysDiff)%360
    mdia=(2*(939.6+61.1*math.cos(math.radians(gm))))
    shdia=(2*(2545.4+228.9*math.cos(math.radians(gm))-16.4*math.cos(math.radians(gs))))


    d=1/2*(mdia+shdia)
    ddash=1/2*(mdia-shdia)
    aLOfMoon=beta*204/205

    # Calculating Type of Eclipse
    eclipseType="no"
    if abs(aLOfMoon)<d:
        if abs(aLOfMoon)<ddash:
            eclipseType="Total"
        else:
            eclipseType="Partial"
            
    pramanam=0        
    if eclipseType != "no":
        pramanam=round((d-abs(aLOfMoon))/mdia,2)

    print(date.strftime('Lunar Eclipse Detail of %d %B %Y'))
    print("Eclipse Type : ",eclipseType)
    print("Magnitude of Eclipse : ",pramanam)


    if eclipseType!="no":
        mdm=47435+5163*math.cos(math.radians(gm))+1008*math.cos(math.radians(2*d))+906*math.cos(math.radians(gm-2*d))+351*math.cos(math.radians(2*gm))-190*math.cos(math.radians(2*d))+124*math.cos(math.radians(gm+2*d))+67*math.cos(math.radians(2*d-gs))+37*math.cos(math.radians(gm-2*d+gs))+31*math.cos(math.radians(gm-gs))-27*math.cos(math.radians(d))-27*math.cos(math.radians(gm+gs))
        sdm=3548+119*math.cos(math.radians(gs))-2*math.cos(math.radians(2*gs))
        vrk=(mdm-sdm)/60
        mdot=(vrk*206)//205
        vrch=trueMoonInSec-trueRahuInSec
        if vrch<0:
            vrch+=1296000

        # Calculating Begin,middle and end timings 
        ioo=angleToSec(time)
        if angleToSec((0,0,0)) <= vrch <= angleToSec((90,0,0)) or angleToSec((180,0,0)) <= vrch <= angleToSec((270,0,0)):
            MID=ioo-vinadiToSec((abs(aLOfMoon)*59)/(10*mdot))
        else:
            MID=ioo+vinadiToSec((abs(aLOfMoon)*59)/(10*mdot))
        start=0
        end=0
        if eclipseType=="Partial":
            HDUR=naligaiToSec(math.sqrt(d**2-aLOfMoon**2)/mdot)
            start=MID-HDUR
            end=MID+HDUR
        else:
            THDUR=naligaiToSec(math.sqrt(ddash**2-aLOfMoon**2)/mdot)
            start=MID-THDUR
            end=MID+THDUR

        print("Timings of Eclipse :")
        print("Beginnig of Eclipse : ",printTime(start))
        print("Middle of Eclipse : ",printTime(MID))
        print("End of Eclipse : ",printTime(end))


# Lunar Eclipse Calculator 
def SolarEclipse(date,time,trueMoon,Rahu):
    """
    Calculate Solar Eclipse 
    Parameters :
    date => Date to check for eclipse in (Y,M,D)
    time => Instant of New Moon
    trueMoon => True Longitude of Moon
    trueRahu => True Longitude of Rahu

    Output:
    Eclipse Type
    Eclipse Magnitude
    If Eclipse Occurs
    Eclipse Beginning Time
    Totality Beginning Time
    Eclipse Middle Time
    Totality End Time
    Eclipse End Time
    """
    date=datetime.datetime(*date)
    trueMoonInSec=angleToSec(trueMoon)
    trueRahuInSec=angleToSec(Rahu)

    daysDiff=trueMoonInSec-trueRahuInSec
    gm=(134.9633964+13.06499295*daysDiff)%360
    gs=(357.5291092+0.985600231*daysDiff)%360

    mdia=2*(939.6+61.1*math.cos(math.radians(gm)))
    sdia=2*(961.2+16.1*math.cos(math.radians(gs)))

    par=3447.9+224.4*math.cos(math.radians(gm))
    d=par+(mdia+sdia)/2
    dDash=par+(mdia-sdia)/2

    # # initially 308 but better result 
    diff=trueMoonInSec-trueRahuInSec
    chandraSara=angleToSec((0,308,0))*math.sin(math.radians(diff/3600))
    lOfMoon=(chandraSara*204)//205

    # Calculating Type of Eclipse
    eclipseType="no"
    if abs(lOfMoon)<d:
        if abs(lOfMoon)<d:
            eclipseType="Total"
        else:
            eclipseType="Partial"
            
    pramanam=0        
    if eclipseType != "no":
        pramanam=round((d-abs(lOfMoon))/mdia,2)
    print(date.strftime('Solar Eclipse Detail of %d %B %Y'))
    print("Eclipse Type : ",eclipseType)
    print("Magnitude of Eclipse : ",pramanam)

    if eclipseType!="no":
        mdm=47435+5163*math.cos(math.radians(gm))+1008*math.cos(math.radians(2*d))+906*math.cos(math.radians(gm-2*d))+351*math.cos(math.radians(2*gm))-190*math.cos(math.radians(2*d))+124*math.cos(math.radians(gm+2*d))+67*math.cos(math.radians(2*d-gs))+37*math.cos(math.radians(gm-2*d+gs))+31*math.cos(math.radians(gm-gs))-27*math.cos(math.radians(d))-27*math.cos(math.radians(gm+gs))
        sdm=3548+119*math.cos(math.radians(gs))-2*math.cos(math.radians(2*gs))
        vrk=(mdm-sdm)/60
        mdot=(vrk*206)//205
        vrch=trueMoonInSec-trueRahuInSec
        if vrch<0:
            vrch+=1296000

        # Calculating Begin,middle and end timings 
        ioo=angleToSec(time)
        if angleToSec((0,0,0)) <= vrch <= angleToSec((90,0,0)) or angleToSec((180,0,0)) <= vrch <= angleToSec((270,0,0)):
            MID=ioo-nadiToSec((99*abs(lOfMoon))/(1000*mdot))
        else:
            MID=ioo+nadiToSec((99*abs(lOfMoon))/(1000*mdot))

        HDUR=naligaiToSec(math.sqrt(d**2-lOfMoon**2)/mdot)
        THDUR=naligaiToSec(math.sqrt(dDash**2-lOfMoon**2)/mdot)
        startEclipse=MID-HDUR
        endEclipse=MID+HDUR
        startTotality=MID-THDUR
        endTotality=MID+THDUR

        print("Timings of Solar Eclipse :")
        print("Beginnig of Eclipse : ",printTime(startEclipse))
        if eclipseType=="Total":
            print("Beginnig of Totality : ",printTime(startTotality))
        print("Middle of Eclipse : ",printTime(MID))
        if eclipseType=="Total":
            print("End of Totality : ",printTime(endTotality))
        print("End of Eclipse : ",printTime(endEclipse))
       

AncIndMatAst/AC.py
import sys
import math
import itertools
sys.stdin.reconfigure(encoding='utf-8')
sys.stdout.reconfigure(encoding='utf-8')

varg={
        'क':1,
        'ख':2,
        'ग':3,
        'घ':4,
        'ङ':5,
        'च':6,
        'छ':7,
        'ज':8,
        'झ':9,
        'ञ':10,
        'ट':11,
        'ठ':12,
        'ड':13,
        'ढ':14,
        'ण':15,
        'त':16,
        'थ':17,
        'द':18,
        'ध':19,
        'न':20,
        'प':21,
        'फ':22,
        'ब':23,
        'भ':24,
        'म':25
    }
avarg={
        'य':3,
        'र':4,
        'ल':5,
        'व':6,
        'श':7,
        'ष':8,
        'स':9,
        'ह':10
    }
swar={
        "ा":{
            'varg':0,
            'avarg':1
        },
        'ि':{
            'varg':2,
            'avarg':3
        },
        'ु':{
            'varg':4,
            'avarg':5
        },
        'ृ':{
            'varg':6,
            'avarg':7
        },
        'ॄ':{
            'varg':8,
            'avarg':9
        },
        'ॆ':{
            'varg':10,
            'avarg':11
        },
        'े':{
            'varg':12,
            'avarg':13
        },
        'ो':{
            'varg':14,
            'avarg':15
        },
        'ौ':{
            'varg':16,
            'avarg':17
        }
}
def decode(sabd):
    i=0
    l=len(sabd)
    value=0
    while i < l :
        schwa=sabd[i]
        if(i!=l-1):
            # if half word 
            if(sabd[i+1]=='्'):
                matra=sabd[i+3]
                i=i+1
            elif(sabd[i+1] in swar):
                matra=sabd[i+1]
                i=i+1
            else:
                matra="ा"
        else:
            matra="ा"
        i=i+1
        type_of_schwa=""
        if(schwa in varg):
            scahwa_val=varg.get(schwa)
            type_of_schwa="varg"
        else:
            scahwa_val=avarg.get(schwa)
            type_of_schwa="avarg"
        value+=scahwa_val*math.pow(10,swar.get(matra).get(type_of_schwa))

    return int(value)
def encode(num):
    aksharList=[]
    posval=0
    while(num!=0):
        skipCnt=1
        fposval=posval
        d=num%10
        num=num//10
        # if number is less than 25 take two digits 
        prevd=num%10
        if posval%2==0:
            if(prevd*10+d<=25):
                d=prevd*10+d
                num=num//10
                skipCnt+=1
        else:
            if(d == 1 or d==2 ):
                d=d*10
                fposval=posval-1
        akshar=""
        # if digit is not zero then only akshar will be generated
        if(d!=0):
            # case varg as pos val is even 
            if(fposval%2==0):
                for key,val in varg.items():
                    if(val==d):
                        akshar+=key
                        break
                if(fposval!=0):
                    for matra,value in swar.items():
                        if value.get("varg")==fposval:
                            akshar+=matra
                            break
            # case avarg as pos val is odd 
            else:
                for key,val in avarg.items():
                    if(val==d):
                        akshar+=key
                        break
                if(fposval!=1):
                    for matra,value in swar.items():
                        if value.get("avarg")==fposval:
                            akshar+=matra
                            break
            if(akshar!=""):
                aksharList.append(akshar)
        posval=posval+skipCnt
    # As a number have multiple Aryabhatiya sabd generating all prmutation if words
    allSabdList=list(itertools.permutations(aksharList))
    # Converting tuple of akshars to string type 
    for i in range(len(allSabdList)):
        sabd=""
        for akshar in allSabdList[i]:
            sabd+=akshar
        allSabdList[i]=sabd
        
    return ({
        "sabd":allSabdList[0],
        "length":len(allSabdList),
        "allSabd":allSabdList
    })

